{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4039845a-bbd4-4c24-b887-f978469d7ee6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, t\n",
      "\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: BiocIO\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n",
      "\n",
      "Attaching package: ‘rtracklayer’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocIO’:\n",
      "\n",
      "    FileForFormat\n",
      "\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'Matrix'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────────────────────────────────────────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32m%within%()\u001b[39m masks \u001b[34mIRanges\u001b[39m::%within%()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcollapse()\u001b[39m     masks \u001b[34mBiostrings\u001b[39m::collapse(), \u001b[34mIRanges\u001b[39m::collapse()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcombine()\u001b[39m      masks \u001b[34mBiobase\u001b[39m::combine(), \u001b[34mBiocGenerics\u001b[39m::combine()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mcompact()\u001b[39m      masks \u001b[34mXVector\u001b[39m::compact()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mdesc()\u001b[39m         masks \u001b[34mIRanges\u001b[39m::desc()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m       masks \u001b[34mMatrix\u001b[39m::expand(), \u001b[34mS4Vectors\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m       masks \u001b[34mensembldb\u001b[39m::filter(), \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m        masks \u001b[34mS4Vectors\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m          masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mpack()\u001b[39m         masks \u001b[34mMatrix\u001b[39m::pack()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggplot2\u001b[39m::\u001b[32mPosition()\u001b[39m   masks \u001b[34mBiocGenerics\u001b[39m::Position(), \u001b[34mbase\u001b[39m::Position()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mreduce()\u001b[39m       masks \u001b[34mGenomicRanges\u001b[39m::reduce(), \u001b[34mIRanges\u001b[39m::reduce(), \u001b[34mSignac\u001b[39m::reduce()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond()\u001b[39m   masks \u001b[34mS4Vectors\u001b[39m::second()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond<-()\u001b[39m masks \u001b[34mS4Vectors\u001b[39m::second<-()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mselect()\u001b[39m       masks \u001b[34mensembldb\u001b[39m::select(), \u001b[34mAnnotationDbi\u001b[39m::select()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mslice()\u001b[39m        masks \u001b[34mXVector\u001b[39m::slice(), \u001b[34mIRanges\u001b[39m::slice()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32munpack()\u001b[39m       masks \u001b[34mMatrix\u001b[39m::unpack()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n"
     ]
    }
   ],
   "source": [
    "library(Signac)\n",
    "library(Seurat)\n",
    "library(GenomeInfoDb)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "library(EnsDb.Mmusculus.v79)\n",
    "library(ggplot2)\n",
    "library(patchwork)\n",
    "library(Matrix)\n",
    "library(GenomicRanges)\n",
    "library(tidyverse)\n",
    "set.seed(1234)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "10815ebe-f491-4d4f-a623-4fc6209dce43",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#\n",
    "LoadData <- function(data_path=NULL, assay='ATAC'){\n",
    "    if (assay=='ATAC'){\n",
    "        atac_assay=readMM(paste0(data_path,'/matrix.mtx'))\n",
    "        barcodes=readLines(paste0(data_path,'/barcodes.tsv'))\n",
    "        peaks=readLines(paste0(data_path,'/peak.bed'))\n",
    "        row.names(atac_assay)=peaks\n",
    "        colnames(atac_assay)=barcodes\n",
    "        #\n",
    "        return (atac_assay)\n",
    "    }\n",
    "    if (assay=='RNA'){\n",
    "        rna_assay=readMM(paste0(data_path,'/matrix.mtx'))\n",
    "        barcodes=readLines(paste0(data_path,'/barcodes.tsv'))\n",
    "        features=readLines(paste0(data_path,'/genes.tsv'))\n",
    "        row.names(rna_assay)=features\n",
    "        colnames(rna_assay)=barcodes\n",
    "        #\n",
    "        return (rna_assay)\n",
    "    }\n",
    "    if (assay=='GeneScore'){\n",
    "        rna_assay=readMM(paste0(data_path,'/matrix.mtx'))\n",
    "        barcodes=readLines(paste0(data_path,'/barcodes.tsv'))\n",
    "        features=readLines(paste0(data_path,'/genes.tsv'))\n",
    "        row.names(rna_assay)=features\n",
    "        colnames(rna_assay)=barcodes\n",
    "        #\n",
    "        return (rna_assay)\n",
    "    }\n",
    "}\n",
    "#\n",
    "Processing_peaks <- function(peak_file=NULL, sep = c(\":\", \"-\"), remove_genome='mm10'){\n",
    "    # loading\n",
    "    peaks <- readLines(peak_file)\n",
    "    \n",
    "    # processing\n",
    "    if (remove_genome=='mm10'){\n",
    "        peaks=peaks[!grepl(paste0(c('mm10'), collapse = \"|\"), peaks)]\n",
    "        peaks=gsub('GRCh38_','',peaks)\n",
    "    }\n",
    "    if (remove_genome=='hg19'){\n",
    "        peaks=peaks[!grepl(paste0(c('GRCh38'), collapse = \"|\"), peaks)]\n",
    "        peaks=gsub('mm10_','',peaks)\n",
    "    }\n",
    "    \n",
    "    # convert to genomic ranges\n",
    "    gr.peaks <- StringToGRanges(regions = peaks, sep = sep)\n",
    "    # standard peaks\n",
    "    gr.peaks <- Standard_peaks(gr.peaks)\n",
    "    return (gr.peaks)\n",
    "}\n",
    "#\n",
    "Processing_counts <- function(counts=NULL, remove_genome='mm10'){\n",
    "    if (remove_genome=='mm10'){\n",
    "        peaks=row.names(counts)\n",
    "        peaks=peaks[!grepl(paste0(c('mm10'), collapse = \"|\"), peaks)]\n",
    "        counts=counts[peaks,]\n",
    "        #\n",
    "        peaks=row.names(counts)\n",
    "        peaks=gsub('hg19_','',peaks)\n",
    "        row.names(counts)=peaks\n",
    "        return (counts)\n",
    "    }\n",
    "    ###\n",
    "    if (remove_genome=='hg19'){\n",
    "        peaks=row.names(counts)\n",
    "        peaks=peaks[!grepl(paste0(c('hg19'), collapse = \"|\"), peaks)]\n",
    "        counts=counts[peaks,]\n",
    "        #\n",
    "        peaks=row.names(counts)\n",
    "        peaks=gsub('mm10_','',peaks)\n",
    "        row.names(counts)=peaks\n",
    "        return (counts)\n",
    "    }\n",
    "}\n",
    "Standard_peaks <- function (rowRanges=rowRanges, annotation='hg38'){\n",
    "    if (annotation=='hg19'){main.chroms <- standardChromosomes(BSgenome.Hsapiens.UCSC.hg19)}\n",
    "    if (annotation=='hg38'){main.chroms <- standardChromosomes(BSgenome.Hsapiens.UCSC.hg38)}\n",
    "    if (annotation=='mm10'){main.chroms <- standardChromosomes(BSgenome.Mmusculus.UCSC.mm10)}\n",
    "    keep.peaks <- which(as.character(seqnames(rowRanges)) %in% main.chroms)\n",
    "    #keep.peaks <- which(as.character(seqnames(rowRanges)) %in% c('chr1'))\n",
    "    rowRanges <- rowRanges[keep.peaks]\n",
    "    \n",
    "    #peaks=as.data.frame(rowRanges)[,c(1,2,3)]\n",
    "    #peaks=peaks[keep.peaks,]\n",
    "    #rowRanges <- makeGRangesFromDataFrame(peaks)\n",
    "    \n",
    "    return (rowRanges)\n",
    "}\n",
    "Create_scRNA_object <- function(data.dir=NULL, rna.assay=NULL){\n",
    "    #\n",
    "    if (is.null(data.dir)){\n",
    "        if (is.null(rna.assay)){stop('Please provide the RNA expression matrix')}\n",
    "    }\n",
    "    else{\n",
    "        rna.assay=Read10X(data.dir = data.dir, gene.column = 1)\n",
    "        #rna.assay=Read10X(data.dir = data.dir, gene.column = 2)\n",
    "    }\n",
    "    #\n",
    "    proj <- CreateSeuratObject(counts = rna.assay)\n",
    "    #\n",
    "    return(proj)\n",
    "}\n",
    "Create_scATAC_object <- function(counts=NULL,fragment_file=NULL){\n",
    "    #\n",
    "    atac.assay <- CreateChromatinAssay(counts, fragments = fragment_file)\n",
    "    atac.obj <- CreateSeuratObject(atac.assay, assay = \"ATAC\")\n",
    "    #\n",
    "    return (atac.obj)\n",
    "}\n",
    "# Do annotation for seurat object\n",
    "annotate_ATAC_obj <- function(proj=proj, annotation='hg38'){\n",
    "    #\n",
    "    if (annotation=='hg38'){\n",
    "        hg38 = \"/jdfssz1/ST_SUPERCELLS/PUB/scRNA/pipeline/v3.1.5/database/Human/genes.gtf\"\n",
    "        gtf <- rtracklayer::import(hg38)\n",
    "        gene.coords <- gtf[gtf$type == 'gene']\n",
    "        seqlevelsStyle(gene.coords) <- 'UCSC'\n",
    "        gene.coords <- keepStandardChromosomes(gene.coords, pruning.mode = 'coarse')\n",
    "        gene.coords$gene_biotype <- gene.coords$gene_type\n",
    "    }\n",
    "    #\n",
    "    if (annotation=='mm10'){\n",
    "        mm10 = \"/jdfssz1/ST_SUPERCELLS/PUB/scRNA/pipeline/v3.1.5/database/Mouse/genes.gtf\"\n",
    "        gtf <- rtracklayer::import(mm10)\n",
    "        gene.coords <- gtf[gtf$type == 'gene']\n",
    "        seqlevelsStyle(gene.coords) <- 'UCSC'\n",
    "        gene.coords <- keepStandardChromosomes(gene.coords, pruning.mode = 'coarse')\n",
    "        gene.coords$gene_biotype <- gene.coords$gene_type\n",
    "    }\n",
    "    # add the gene information to the object\n",
    "    Annotation(proj) <- gene.coords\n",
    "    #\n",
    "    return (proj)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e907a2a-b445-4604-8f96-0f87a556f20a",
   "metadata": {
    "tags": []
   },
   "source": [
    "# nFeature/nCounts per cell"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60d4af20-95b6-485f-bda2-35f5e0e5465f",
   "metadata": {},
   "source": [
    "## ATAC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8ac5c11-7c4f-464d-95f2-1fa1ef4e36eb",
   "metadata": {},
   "source": [
    "### single sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6706e4df-cd43-414c-b8ba-0454c4b1baaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path='/hwfssz1/ST_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/cellline_ISSAAC_processed/IL10/ATAC/downsampled/'\n",
    "sample_names=c()\n",
    "names=c('HT-scCAT-seq_EU11_1')\n",
    "human=c('HEK293T')\n",
    "mouse=c('NIH3T3')\n",
    "counts_files_human=c(paste0(data_path, human))\n",
    "counts_files_mouse=c(paste0(data_path, mouse))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1348f486-91c6-4216-933d-5c40dea0aab5",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (i in 1:length(counts_files_human)){\n",
    "    counts <- LoadData(data_path=counts_files_human)\n",
    "    counts <- Processing_counts(counts, remove_genome='mm10')\n",
    "    seurat.obj=Create_scATAC_object(counts=counts, fragment_file=NULL)\n",
    "    meta=seurat.obj@meta.data[, c('nCount_ATAC', 'nFeature_ATAC')]\n",
    "    colnames(meta)=c('nCounts', 'nFeatures')\n",
    "    #\n",
    "    output_dir='/hwfssz1/ST_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/cellline_ISSAAC_processed/IL10/ATAC/downsampled/'\n",
    "    output_path=paste0(output_dir, human)\n",
    "    if (!dir.exists(output_path)){dir.create(output_path)}\n",
    "    write.csv(meta, paste0(output_path,'/metrics.csv'), quote=FALSE)\n",
    "}\n",
    "for (i in 1:length(counts_files_mouse)){\n",
    "    counts <- LoadData(data_path=counts_files_mouse)\n",
    "    counts <- Processing_counts(counts, remove_genome='hg19')\n",
    "    seurat.obj=Create_scATAC_object(counts=counts, fragment_file=NULL)\n",
    "    meta=seurat.obj@meta.data[, c('nCount_ATAC', 'nFeature_ATAC')]\n",
    "    colnames(meta)=c('nCounts', 'nFeatures')\n",
    "    #\n",
    "    output_dir='/hwfssz1/ST_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/cellline_ISSAAC_processed/IL10/ATAC/downsampled/'\n",
    "    output_path=paste0(output_dir, mouse)\n",
    "    if (!dir.exists(output_path)){dir.create(output_path)}\n",
    "    write.csv(meta, paste0(output_path,'/metrics.csv'), quote=FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9498ceb8-d78f-4537-87e0-63674eae86b0",
   "metadata": {},
   "source": [
    "### multi-samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c918c0e2-8e29-4cc4-b443-4fae3b381815",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path='/jdfssz1/ST_SUPERCELLS/P21Z10200N0090/Automated/USER/zhangzhongjin/project/celline/sample/all_tech/'\n",
    "sample_names=c('CAT_EU11_1', 'CAT_EU14', 'CAT_FL16_1')\n",
    "names=c('HT-scCAT-seq_EU11_1', 'HT-scCAT-seq_EU14', 'HT-scCAT-seq_FL16_1')\n",
    "human=c('HEK293T')\n",
    "mouse=c('NIH3T3')\n",
    "ATAC=c('ATAC')\n",
    "#\n",
    "counts_files_human=c()\n",
    "counts_files_mouse=c()\n",
    "for (i in 1:length(sample_names)){\n",
    "    counts_files_human=c(counts_files_human, paste0(data_path, sample_names[i],\"/\", ATAC,\"/\", human))\n",
    "    counts_files_mouse=c(counts_files_mouse, paste0(data_path, sample_names[i],\"/\", ATAC,\"/\", mouse))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "65864b3e-1912-4e99-b696-291d7ae72502",
   "metadata": {},
   "outputs": [],
   "source": [
    "names_human=c()\n",
    "names_mouse=c()\n",
    "for (i in 1:length(names)){\n",
    "names_human[i]=c(paste0(names[i], '_', human))\n",
    "names_mouse[i]=c(paste0(names[i], '_', mouse))\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b720261c-3ebf-4a0e-b156-d82e76d3b4db",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (i in 1:length(counts_files_human)){\n",
    "    counts <- LoadData(data_path=counts_files_human[i])\n",
    "    counts <- Processing_counts(counts, remove_genome='mm10')\n",
    "    seurat.obj=Create_scATAC_object(counts=counts, fragment_file=NULL)\n",
    "    meta=seurat.obj@meta.data[, c('nCount_ATAC', 'nFeature_ATAC')]\n",
    "    colnames(meta)=c('nCounts', 'nFeatures')\n",
    "    #\n",
    "    output_dir='/jdfssz1/ST_SUPERCELLS/P21Z10200N0090/Automated/USER/zhangzhongjin/project/celline/sample/compare_data/ATAC/'\n",
    "    output_path=paste0(output_dir, names_human[i])\n",
    "    if (!dir.exists(output_path)){dir.create(output_path)}\n",
    "    write.csv(meta, paste0(output_path,'/metrics.csv'), quote=FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "97948986-6ec6-4c8e-aeaa-82393b7092d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (i in 1:length(counts_files_mouse)){\n",
    "    counts <- LoadData(data_path=counts_files_mouse[i])\n",
    "    counts <- Processing_counts(counts, remove_genome='hg19')\n",
    "    seurat.obj=Create_scATAC_object(counts=counts, fragment_file=NULL)\n",
    "    meta=seurat.obj@meta.data[, c('nCount_ATAC', 'nFeature_ATAC')]\n",
    "    colnames(meta)=c('nCounts', 'nFeatures')\n",
    "    #\n",
    "    output_dir='/jdfssz1/ST_SUPERCELLS/P21Z10200N0090/Automated/USER/zhangzhongjin/project/celline/sample/compare_data/ATAC/'\n",
    "    output_path=paste0(output_dir, names_mouse[i])\n",
    "    if (!dir.exists(output_path)){dir.create(output_path)}\n",
    "    write.csv(meta, paste0(output_path,'/metrics.csv'), quote=FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1942e582-5dcb-49f4-b85c-d4f29cd20163",
   "metadata": {},
   "source": [
    "## RNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f16e9454-f010-485e-9d9a-b2488ab77d34",
   "metadata": {},
   "source": [
    "### single sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f4eded81-1d5e-4bd2-90a5-c0ae600a9a01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/hwfssz1/ST_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/cellline_ISSAAC_processed/ISSAAC/RNA/downsampled/HEK293T'"
      ],
      "text/latex": [
       "'/hwfssz1/ST\\_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/cellline\\_ISSAAC\\_processed/ISSAAC/RNA/downsampled/HEK293T'"
      ],
      "text/markdown": [
       "'/hwfssz1/ST_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/cellline_ISSAAC_processed/ISSAAC/RNA/downsampled/HEK293T'"
      ],
      "text/plain": [
       "[1] \"/hwfssz1/ST_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/cellline_ISSAAC_processed/ISSAAC/RNA/downsampled/HEK293T\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "'/hwfssz1/ST_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/cellline_ISSAAC_processed/ISSAAC/RNA/downsampled/NIH3T3'"
      ],
      "text/latex": [
       "'/hwfssz1/ST\\_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/cellline\\_ISSAAC\\_processed/ISSAAC/RNA/downsampled/NIH3T3'"
      ],
      "text/markdown": [
       "'/hwfssz1/ST_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/cellline_ISSAAC_processed/ISSAAC/RNA/downsampled/NIH3T3'"
      ],
      "text/plain": [
       "[1] \"/hwfssz1/ST_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/cellline_ISSAAC_processed/ISSAAC/RNA/downsampled/NIH3T3\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data_path='/hwfssz1/ST_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/cellline_ISSAAC_processed/ISSAAC/RNA/downsampled/'\n",
    "sample_names=c()\n",
    "names=c('HT-scCAT-seq_IL9')\n",
    "human=c('HEK293T')\n",
    "mouse=c('NIH3T3')\n",
    "counts_files_human=c(paste0(data_path, human))\n",
    "counts_files_mouse=c(paste0(data_path, mouse))\n",
    "counts_files_human\n",
    "counts_files_mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e1c05a43-a9e0-47a4-b042-9a8713b42ad9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Feature names cannot have underscores ('_'), replacing with dashes ('-')\"\n",
      "Warning message:\n",
      "\"Feature names cannot have underscores ('_'), replacing with dashes ('-')\"\n"
     ]
    }
   ],
   "source": [
    "# HEK293T\n",
    "counts <- LoadData(data_path=counts_files_human, assay='RNA')\n",
    "counts <- as(counts, 'CsparseMatrix')\n",
    "seurat.obj=CreateSeuratObject(counts=counts)\n",
    "meta=seurat.obj@meta.data[, c('nCount_RNA', 'nFeature_RNA')]\n",
    "colnames(meta)=c('nCounts', 'nFeatures')\n",
    "output_dir='/hwfssz1/ST_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/cellline_ISSAAC_processed/ISSAAC/RNA/downsampled/HEK293T'\n",
    "write.csv(meta, paste0(output_dir,'/metrics.csv'), quote=FALSE)\n",
    "# NIH3T3\n",
    "counts <- LoadData(data_path=counts_files_mouse, assay='RNA')\n",
    "counts <- as(counts, 'CsparseMatrix')\n",
    "seurat.obj=CreateSeuratObject(counts=counts)\n",
    "meta=seurat.obj@meta.data[, c('nCount_RNA', 'nFeature_RNA')]\n",
    "colnames(meta)=c('nCounts', 'nFeatures')\n",
    "output_dir='/hwfssz1/ST_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/cellline_ISSAAC_processed/ISSAAC/RNA/downsampled/NIH3T3'\n",
    "write.csv(meta, paste0(output_dir,'/metrics.csv'), quote=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "038aaf79-318f-42ec-be55-a0ed6d17b0de",
   "metadata": {},
   "source": [
    "### multi-samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "243e4b22-eb78-4254-8c00-a04fb61aa155",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path='/hwfssz1/ST_SUPERCELLS/P21Z10200N0090/zhangzhongjin/celline/sample/cortex/EL13/RNA/merged/'\n",
    "sample_names=c('CAT_EU11_1', 'CAT_EU14', 'CAT_FL16_1')\n",
    "names=c('HT-scCAT-seq_EU11_1', 'HT-scCAT-seq_EU14', 'HT-scCAT-seq_FL16_1')\n",
    "human=c('HEK293T')\n",
    "mouse=c('NIH3T3')\n",
    "RNA=c('RNA')\n",
    "#\n",
    "counts_files_human=c()\n",
    "counts_files_mouse=c()\n",
    "for (i in 1:length(sample_names)){\n",
    "    counts_files_human=c(counts_files_human, paste0(data_path, sample_names[i],\"/\", RNA,\"/\", human))\n",
    "    counts_files_mouse=c(counts_files_mouse, paste0(data_path, sample_names[i],\"/\", RNA,\"/\", mouse))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0e78460a-77a1-4ff8-8586-0d1fdb0d4e1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (i in 1:length(counts_files_human)){\n",
    "    counts <- LoadData(data_path=counts_files_human[i], assay='RNA')\n",
    "    counts <- as(counts, 'CsparseMatrix')\n",
    "    seurat.obj=CreateSeuratObject(counts=counts)\n",
    "    meta=seurat.obj@meta.data[, c('nCount_RNA', 'nFeature_RNA')]\n",
    "    colnames(meta)=c('nCounts', 'nFeatures')\n",
    "    #\n",
    "    output_dir='/jdfssz1/ST_SUPERCELLS/P21Z10200N0090/Automated/USER/zhangzhongjin/project/celline/sample/compare_data/RNA/'\n",
    "    output_path=paste0(output_dir, names_human[i])\n",
    "    if (!dir.exists(output_path)){dir.create(output_path)}\n",
    "    write.csv(meta, paste0(output_path,'/metrics.csv'), quote=FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "952a55c1-8e94-40bc-aa80-0e6047b5951d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Feature names cannot have underscores ('_'), replacing with dashes ('-')\"\n",
      "Warning message:\n",
      "\"Feature names cannot have underscores ('_'), replacing with dashes ('-')\"\n",
      "Warning message:\n",
      "\"Feature names cannot have underscores ('_'), replacing with dashes ('-')\"\n",
      "Warning message:\n",
      "\"Feature names cannot have underscores ('_'), replacing with dashes ('-')\"\n",
      "Warning message:\n",
      "\"Feature names cannot have underscores ('_'), replacing with dashes ('-')\"\n",
      "Warning message:\n",
      "\"Feature names cannot have underscores ('_'), replacing with dashes ('-')\"\n"
     ]
    }
   ],
   "source": [
    "for (i in 1:length(counts_files_mouse)){\n",
    "    counts <- LoadData(data_path=counts_files_mouse[i], assay='RNA')\n",
    "    counts <- as(counts, 'CsparseMatrix')\n",
    "    seurat.obj=CreateSeuratObject(counts=counts)\n",
    "    meta=seurat.obj@meta.data[, c('nCount_RNA', 'nFeature_RNA')]\n",
    "    colnames(meta)=c('nCounts', 'nFeatures')\n",
    "    #\n",
    "    output_dir='/jdfssz1/ST_SUPERCELLS/P21Z10200N0090/Automated/USER/zhangzhongjin/project/celline/sample/compare_data/RNA/'\n",
    "    output_path=paste0(output_dir, names_mouse[i])\n",
    "    if (!dir.exists(output_path)){dir.create(output_path)}\n",
    "    write.csv(meta, paste0(output_path,'/metrics.csv'), quote=FALSE)\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA",
   "language": "R",
   "name": "ir43"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
